### PR TITLE
 Fixed unsafe string dup behaviour.

### PR DESCRIPTION
**Pull request by lee-emqx**

### CODE REVIEW

File: src/supplemental/mqtt/mqtt_msg.c [hunk 1]
```diff
@@ -828,10 +828,12 @@ nni_mqtt_topic_qos_array_create(size_t n)
 
 void
 nni_mqtt_topic_qos_array_set(nni_mqtt_topic_qos *topic_qos, size_t index,
-    const char *topic_name, uint8_t qos, uint8_t nl, uint8_t rap, uint8_t rh)
+    const char *topic_name, uint32_t len, uint8_t qos, uint8_t nl, uint8_t rap, uint8_t rh)
 {
-	topic_qos[index].topic.buf       = (uint8_t *) nni_strdup(topic_name);
-	topic_qos[index].topic.length    = (uint32_t) strlen(topic_name);
+	topic_qos[index].topic.buf       = (uint8_t *) nni_alloc((len+1) * sizeof(uint8_t));
```
Comment (JaylinYu) [inline src/supplemental/mqtt/mqtt_msg.c]: no \0

**Comment by JaylinYu** [comment]
@alvin1221 plz fix related unsafe strcmp

**Review by JaylinYu — APPROVED** [review]

**Review by JaylinYu — APPROVED** [review]
alvin thank lee

File: src/supplemental/mqtt/mqtt_msg.c [hunk 2]
```diff
@@ -828,10 +828,11 @@ nni_mqtt_topic_qos_array_create(size_t n)
 
 void
 nni_mqtt_topic_qos_array_set(nni_mqtt_topic_qos *topic_qos, size_t index,
-    const char *topic_name, uint8_t qos, uint8_t nl, uint8_t rap, uint8_t rh)
+    const char *topic_name, uint32_t len, uint8_t qos, uint8_t nl, uint8_t rap, uint8_t rh)
 {
-	topic_qos[index].topic.buf       = (uint8_t *) nni_strdup(topic_name);
-	topic_qos[index].topic.length    = (uint32_t) strlen(topic_name);
+	topic_qos[index].topic.buf       = (uint8_t *) nni_alloc(len * sizeof(uint8_t));
```
Comment (wanghaEMQ) [inline src/supplemental/mqtt/mqtt_msg.c:833]: Why not just use function strndup()?

**Pull request merged**